Creating API contract with Design-first or Code-first Approach

cpeeyush opened the pull request:
# Pull Request Template

## What are you trying to address

- Details on different approaches of creating an Open API contract.
- Addresses #660 

## Description of new changes

- As now days development teams widely use API description formats like Swagger (OpenAPI) to describe an API contracts, which brings a question whether a dev team should design an API contract first or write code first and generate api contract out of it. This PR adds a section in playbook to describe both API development approaches with benefits and points to consider.

## For all pull requests

- [x] Link to the issue you are solving (so it gets closed)
- [x] Label the pull request with the appropriate area(s)
- [x] Assign potential reviewers (you may also want to contact them on Teams to ensure timely reviews)
- [x] Assign the project - Engineering Playbook Backlog
- [x] Assign the pull request to yourself

charleszipp commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+
+Important Points to consider:
+
+- If API requirements can changes often during initial development phase than Design-First approach may not be a good fit as this will introduce an overhead to always first update & maintain API contract definitions again & again.
```
```suggestion
- If API requirements changes often during initial development phase, than a Design-First approach may not be a good fit as this will introduce an overhead to always first update & maintain API contract definitions again & again.
```

charleszipp commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+Important Points to consider:
+
+- If API requirements can changes often during initial development phase than Design-First approach may not be a good fit as this will introduce an overhead to always first update & maintain API contract definitions again & again.
+- It might be worthwhile to first try out your platform specific code generator because it is possible that a particular platform specific code generator might not be able to generate a flexible & maintainable implementation of actual code. For instance If your web framework requires annotations to be present on your controller classes (e.g. for API versioning or authentication), make sure that t ...
```
> _...might not be able to generate a flexible & maintainable implementation..._

This point is very subjective. Can we be more concrete or objective about what we are guiding people to consider? For example, should we instead advise the audience to consider how much additional work will be required for the generated code to meet done-done criteria?

cpeeyush replied:
Thanks for the review. 
Updated the section as per your suggestion to bring more clarity. Pleas take a look again. Thanks.

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+# Creating API Contract
+
+As different development teams exposes APIs to access various REST based services, its important to have an API contract to share between producer and consumers of APIs. [Open API](https://www.openapis.org/) format is one of the most popular API description format. This Open API document can be produced in two ways:
```
NIT: 
```suggestion
As different development teams expose APIs to access various REST based services, it's important to have an API contract to share between the producer and consumers of APIs. [Open API](https://www.openapis.org/) format is one of the most popular API description format. This Open API document can be produced in two ways:
```

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+As different development teams exposes APIs to access various REST based services, its important to have an API contract to share between producer and consumers of APIs. [Open API](https://www.openapis.org/) format is one of the most popular API description format. This Open API document can be produced in two ways:
+
+- Design-First - Team starts developing APIs by first describing API designs as an Open API document and later generates server side boilerplate code with the help of this document.
+- Code-First - Team starts writing the server side api interface code e.g. controllers, DTOs etc. and later generates and Open API document from it.
```
```suggestion
- Code-First - Team starts writing the server side API interface code e.g. controllers, DTOs etc. and later generates and Open API document from it.
```

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+
+## Design-First Approach
+
+A Design-First approach means that APIs are treated as "first-class citizens" and everything about a project revolves around the idea that at the end these APIs will be consumed by clients. So based on the business requirements api development team first start describing API designs as an Open API document and collaborate with the stakeholders to gather feedback.
```
NIT: have noticed that in some places API is capitalized but this is not consistent throughout the document, suggest reviewing to make it consistent everywhere.

cpeeyush replied:
Updated it. Thank you.

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+
+A Design-First approach means that APIs are treated as "first-class citizens" and everything about a project revolves around the idea that at the end these APIs will be consumed by clients. So based on the business requirements api development team first start describing API designs as an Open API document and collaborate with the stakeholders to gather feedback.
+
+This approach is quite useful if a project is about developing externally exposed set of APIs which will be consumed by partners. As here we first agree upon an API contract (Open API document) that's why expectations on api producer and consumer side remains strictly clear and both side teams can work in parallel as per the pre-agreed API design.
```
```suggestion
This approach is quite useful if a project is about developing externally exposed set of APIs which will be consumed by partners. In this approach, we first agree upon an API contract (Open API document) creating clear expectations on both api producer and consumer sides so both teams can begin work in parallel as per the pre-agreed API design.
```

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+Key Benefits of this approach:
+
+- Early API design feedback.
+- Expectations remains clear on consumer & producer side as both agreed upon an API contract.
```
```suggestion
- Clearly established expectations for both consumer & producer as both have agreed upon an API contract.
```

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+- Early API design feedback.
+- Expectations remains clear on consumer & producer side as both agreed upon an API contract.
+- Development teams can work in parallel.
+- Testing team can use API contracts to do an early test even before a business logic is in place. By looking at different models, paths, attributes and other aspects of the API testing can provide their input which can be very valuable.
```
```suggestion
- Testing team can use API contracts to write early tests even before business logic is in place. By looking at different models, paths, attributes and other aspects of the API testing can provide their input which can be very valuable.
```

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+
+Important Points to consider:
+
+- If API requirements changes often during initial development phase, than a Design-First approach may not be a good fit as this will introduce an overhead to always first update & maintain API contract definitions again & again.
```
```suggestion
- If API requirements changes often during initial development phase, than a Design-First approach may not be a good fit as this will introduce additional overhead, requiring repeated updates & maintenance to the API contract definitions.
```

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+
+A Code-First approach means that development teams first implements server side api interface code e.g. controllers, DTOs etc. and than generates API contract definitions out of it. In current times this approach is more widely popular within developer community than Design-First Approach.
+
+This approach has advantage if team wants to quickly implement APIs and also provides flexibility to react very quickly on an unexpected API requirement change.
```
```suggestion
This approach has the advantages of allowing the team to quickly implement APIs and also providing the flexibility to react very quickly to any unexpected API requirement changes.
```

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+
+Important Points to consider:
+
+- A generated Open API definition can be outdated, so its important to have automated checks to avoid this otherwise generated client SDKs will be out of sync and may cause issues for API consumers.
```
```suggestion
- A generated Open API definition can become outdated, so its important to have automated checks to avoid this otherwise generated client SDKs will be out of sync and may cause issues for API consumers.
```

fnocera commented on docs/design/design-patterns/rest-api-design-guidance/creating-api-contract.md:
```diff
@@ -0,0 +1,56 @@
+# Creating API Contract
```
```suggestion
# Creating API Contracts
```

fnocera reviewed (commented):
This is great, thanks for the contribution @cpeeyush. 

I would suggest adding a couple of links to any good external resources (if there are any) that you would recommend to someone wanting to get a bit more detail on these concepts. 

#642 has been closed so I recommend adding this content to a section within this page https://github.com/microsoft/code-with-engineering-playbook/tree/main/docs/design/design-patterns/rest-api-design-guidance (perhaps putting it under a header such as "Creating an API Contract" or "Creating API Contracts"). If you want to keep the content as it's own page then still add a section with the same heading in the rest-api-design-guidance page and write a few sentences saying that there is a Code-First or a Design-First approach and then link to the doc you have here. 

Once these changes are made I'll change my review to approve. Thanks again!

cpeeyush commented:
> This is great, thanks for the contribution @cpeeyush.
> 
> I would suggest adding a couple of links to any good external resources (if there are any) that you would recommend to someone wanting to get a bit more detail on these concepts.
> 
> #642 has been closed so I recommend adding this content to a section within this page https://github.com/microsoft/code-with-engineering-playbook/tree/main/docs/design/design-patterns/rest-api-design-guidance (perhaps putting it under a header such as "Creating an API Contract" or "Creating API Contracts"). If you want to keep the content as it's own page then still add a section with the same heading in the rest-api-design-guidance page and write a few sentences saying that there is a Code-First or a Design-First approach and then link to the doc you have here.
> 
> Once these changes are made I'll change my review to approve. Thanks again!

@fnocera Thank you for your review. As suggested I have now moved _Creating API Contracts_ as a separate section under _REST API Design Guidance_ document and also added a official openAPI best practices link around OpenAPI contracts. In general on internet couple of standard blog posts are available around this concept but I am not adding them as further resources because they are particularly opiniated on one of the approach.

Please take a look now and let me know if any further change needed. Thank you.

fnocera reviewed (approved):
Thanks for addressing the PR comments @cpeeyush and thanks for the contribution!
